src/cocotb/handle.py

```diff
@@ -181,130 +223,245 @@ def _discover_all(self) -> None:
+        """Compute the path string of a child object at the given key.
 
-        new_handle = self._handle.get_handle_by_name(name)
+        Args:
+            key: The key of the child object.
 
-        if not new_handle:
-            self._invalid_sub_handles.add(name)
-            return None
+        Returns:
+            A path string of the child object at the a given key.
+        """
 
-        sub_handle = SimHandle(new_handle, self._child_path(name))
-        self._sub_handles[name] = sub_handle
-        return sub_handle
+    @abstractmethod
+    def _sub_handle_key(self, name: str) -> KeyType:
+        """Translate a discovered child object name into a key.
 
-    def __setattr__(self, name, value):
-        """Provide transparent access to signals via the hierarchy.
+        Args:
+            name: The GPI name of the child object.
 
-        Slightly hacky version of operator overloading in Python.
+        Returns:
+            A unique key for the child object.
 
-        Raise an :exc:`AttributeError` if users attempt to create new members which
-        don't exist in the design.
+        Raises:
+            ValueError: if unable to translate handle to a valid _sub_handle key.
         """
 
+    def __iter__(self) -> Iterable[SimHandleBase]:
+        return iter(self._values())
+
+    def __len__(self) -> int:
+        self._discover_all()
+        return len(self._sub_handles)
+
+    def __dir__(self) -> List[str]:
+        """Permits IPython tab completion and debuggers to work."""
+        self._discover_all()
+        return super().__dir__() + [str(k) for k in self._keys()]
+
+
+class HierarchyObject(HierarchyObjectBase[str]):
+    r"""Named hierarchical scope objects.
+
+    This class is used for named hierarchical structures, such as "generate blocks" or "module"/"entity" instantiations.
+
+    Children under this structure are found by using the name of the child with either the attribute syntax or index syntax.
+    For example, if in your :envvar:`TOPLEVEL` entity/module you have a signal/net named ``count``, you could do either of the following.
+
+    .. code-block:: python3
+
+        dut.count  # attribute syntax
+        dut["count"]  # index syntax
+
+    Attribute syntax is usually shorter and easier to read, and is more common.
+    However, it has limitations:
+
+    - the name cannot start with a number
+    - the name cannot start with a ``_`` character
+    - the name can only contain ASCII letters, numbers, and the ``_`` character
+
+    Any possible name of an object is supported with the index syntax,
+    but it can be more verbose.
+
+    .. note::
+        If you need to access signals/nets that start with ``_``,
+        or use escaped identifier (Verilog) or extended identifier (VHDL) characters,
+        you have to use the index syntax.
+        Accessing escaped/extended identifiers requires enclosing the name with leading and trailing backslashes (``\\``).
```

Comment:
```suggestion
        Accessing escaped/extended identifiers requires enclosing the name
        with leading and trailing double backslashes (``\\``).
```
Now it should be completely clear.